dvc/remote/local.py

```diff
@@ -354,6 +375,29 @@ def make_names(checksum, names):
 
         return dir_status, file_status, dir_paths
 
+    def _indexed_dir_checksums(self, named_cache, remote, dir_md5s):
+        # Validate our index by verifying all indexed .dir checksums
+        # still exist on the remote
+        dir_exists = remote._cache_object_exists(dir_md5s)
+        indexed_dirs = set(remote.index.intersection(dir_md5s))
+        missing_dirs = indexed_dirs.difference(dir_exists)
+        if missing_dirs:
+            logger.debug(
+                "Remote cache missing indexed .dir checksums '{}', "
+                "clearing remote index".format(", ".join(missing_dirs))
+            )
+            remote.index.clear()
+        else:
+            # If .dir checksum exists on the remote, assume indexed
+            # directory contents still exists on the remote
+            for dir_checksum in dir_exists:
+                yield from remote.index.intersection(
+                    named_cache.child_keys(self.scheme, dir_checksum)
+                )
```

Comment:
Safer to assume that only indexed dir contents exists here, rather than assuming that the full dir contents exists.